src/language_features/formatting.rs

```diff
@@ -1,31 +1,77 @@
+use std::collections::HashMap;
+
 use crate::capabilities::{attempt_server_capability, CAPABILITY_FORMATTING};
 use crate::context::*;
 use crate::types::*;
+use crate::util::editor_quote;
+use itertools::Itertools;
 use lsp_types::request::*;
 use lsp_types::*;
 use serde::Deserialize;
 use url::Url;
 
 pub fn text_document_formatting(meta: EditorMeta, params: EditorParams, ctx: &mut Context) {
-    if meta.fifo.is_none() && !attempt_server_capability(ctx, &meta, CAPABILITY_FORMATTING) {
+    let eligible_servers: Vec<_> = ctx
+        .language_servers
+        .iter()
+        .filter(|server| attempt_server_capability(*server, &meta, CAPABILITY_FORMATTING))
+        .filter(|(server_name, _)| {
+            if let Some(fmt_server) = &meta.server {
+                *server_name == fmt_server
+            } else {
+                true
+            }
+        })
+        .collect();
+    if meta.fifo.is_none() && eligible_servers.is_empty() {
+        return;
+    }
+
+    // Ask user to pick which server to use for formatting when multiple options are available.
+    if eligible_servers.len() > 1 {
+        let choices = eligible_servers
+            .into_iter()
+            .map(|(server_name, _)| {
+                let cmd = if meta.fifo.is_some() {
+                    "lsp-formatting-sync"
+                } else {
+                    "lsp-formatting"
+                };
+                let cmd = format!("{} {}", cmd, server_name);
+                format!("{} {}", editor_quote(server_name), editor_quote(&cmd))
+            })
+            .join(" ");
+        ctx.exec(meta, format!("lsp-menu {}", choices));
```

Comment:
I'm not sure about this. Probably the user should decide this earlier.
Also we'll want a non-interactive version for hooks.

The same problem exists for many other requests besides formatting, right?

Need to think about which ones have such conflicts.
An alternative solution is to have a primary server for such ambiguous requests.
I wonder what fellow LSP clients do.

Reply:
Indeed, although rarer for other methods and much more common for formatting, this race condition problem exists.

Note that, if only a single server is used, then there is no prompt anyway, so this would be a breaking change for users.

IMO, having a prompt for picking the formatting server is a good default UX for a multi-server setup (just reminding that the prompt won't be shown to most users, who will still use a single server).

Having priorities configured is also good UX IMO, and I think both can co-exist if the user doesn't want a prompt or if the user doesn't want to set up `lsp-formatting <server>` for languages that use multiple servers.

For users that desire to use `lsp-formatting` and the likes in a non-interactive way with multiple servers, one can inform such commands the desired server (for example `lsp-formatting tsserver`). By passing the server name to the command, there will never be a prompt.

Reply:
ok a prompt sounds like a reasonable default UI.
As for how to force a specific set of servers, I'm not sure if arguments are the best way to do this.
Maybe we should have a `str-list` option that users can set before their request..